web-api/src/personalization/rerank.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright 2023 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::{cmp::Ordering, collections::HashMap, hash::Hash};
+
+use itertools::{izip, Itertools};
+use xayn_ai_coi::{nan_safe_f32_cmp_desc, CoiSystem, UserInterests};
+
+use crate::models::{DocumentId, DocumentTag, PersonalizedDocument};
+
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+#[repr(transparent)]
+pub(super) struct Rank(usize);
+
+impl Rank {
+    pub(super) fn from_index(index: usize) -> Rank {
+        Rank(index + 1)
+    }
+
+    pub(super) fn to_score(self) -> f32 {
+        #![allow(clippy::cast_precision_loss)]
+        1. / self.0 as f32
```

Comment:
Arbitrary but when merging rankings we want to make sure that the score we return still has the characteristics of an score i.e.:

- bigger is better
- it's in `[0;1]`

This doesn't affect the outcome of the algorithm wrt. the resulting order. We just use it to compute the combined score which AFIK is currently not defined by pink. When it will be defined we will change it. But we need something now.